### PR TITLE
ARROW-12199: [R] bindings for stddev, variance

### PR DESCRIPTION
**Pull request by thisisnic**
Adds bindings for stddev and variance, and also uses these kernels in the dplyr expressions when calls to sd/var

### CODE REVIEW

File: r/R/dplyr.R [hunk 1]
```diff
@@ -480,6 +480,18 @@ build_function_list <- function(FUN) {
     between = function(x, left, right) {
       x >= left & x <= right
     },
+    sd = function(x, na.rm = FALSE){
+      if (!na.rm && x$null_count > 0) {
+        return(Scalar$create(NA_real_))
+      }
```
Comment (thisisnic) [inline r/R/dplyr.R:486]: I'm not sure if this is wrong somehow, as when I call the code below, I get the warning `Warning: Expression mass - sd(mass, na.rm = FALSE) not supported in Arrow; pulling data into R`, but yet when I set `na.rm` to TRUE, I don't.
```
Table$create(starwars) %>%
     select(name, mass, species) %>%
     mutate(mass_diff = mass-sd(mass, na.rm = FALSE)) %>%
     collect()
```

File: r/R/compute.R [hunk 2]
```diff
@@ -267,6 +267,25 @@ value_counts <- function(x) {
...
+#' `variance` and `stddev` for Arrow objects
+#'
+#' These functions calculate the variance and standard deviation of Arrow arrays
+#' @param x `Array` or `ChunkedArray`
+#' @param ddof The divisor used in calculations is N - ddof, where N is the number of elements. 
+#' By default, ddof is zero, and population variance or stddev is returned. 
+#' @return A `Scalar` containing the calculated value.
+#' @export
+stddev <- function(x, ddof = 0) {
+  call_function("stddev", x, options = list(ddof = ddof))
```
Comment (nealrichardson) [inline r/R/compute.R:280]: Is there no `na.rm` handling in the Arrow stddev and variance functions? If not, there should be (please JIRA).

File: r/R/dplyr.R (hunk 1 above)
Comment (nealrichardson) [reply r/R/dplyr.R:486]: We don't support aggregations in our dplyr backend yet, so this should never succeed. If `sd()` doesn't cleanly and always error when called on an arrow Expression, we should force it to--see the "fail" handling inside of `arrow_eval` where this is done for `mean`.

File: r/src/compute.cpp [hunk 3]
```diff
@@ -232,7 +232,12 @@ std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(
                                      cpp11::as_cpp<std::string>(options["replacement"]),
                                      max_replacements);
   }
-
+  
+  if (func_name == "variance" || func_name == "stddev") {
```
Comment (nealrichardson) [inline r/src/compute.cpp:236]: TBH this is probably the only code addition we want to keep here.

File: r/R/compute.R [hunk 4]
```diff
@@ -267,6 +267,25 @@ value_counts <- function(x) {
   call_function("value_counts", x)
 }
 
+
+#' `variance` and `stddev` for Arrow objects
```
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/R/compute.R]: Got it, removed now.

File: r/R/dplyr.R (hunk 1 above)
Comment (thisisnic) [reply r/R/dplyr.R]: Done.

**Review by nealrichardson — COMMENTED** [review]
Should we add _a_ test of `call_function("stddev", array, options = list(ddof = something))` just to confirm we're passing the argument? ("no" is an acceptable answer)

**Pull request closed without merging**